File: native/ios/ApollosShell/RustCoreBridge.swift
import Foundation

struct IOSKinematicResult {
    let riskScore: Float
    let shouldCapture: Bool
    let yawDeltaDeg: Float
}

struct IOSDepthHazardResult {
    let detected: Bool
    let positionX: Float
    let confidence: Float
    let distanceCode: UInt8
}

struct IOSEskfSnapshot {
    let sensorHealthScore: Float
    let degraded: Bool
    let localizationUncertaintyM: Float
    let innovationNorm: Float
    let covarianceXx: Float
    let covarianceYy: Float
    let covarianceZz: Float
}

struct IOSVisionOdometryResult {
    let applied: Bool
    let deltaXM: Float
    let deltaYM: Float
    let poseXM: Float
    let poseYM: Float
    let varianceM2: Float
    let opticalFlowScore: Float
    let lateralBias: Float
}

struct IOSEdgeObjectDetection {
    let labelId: UInt32
    let xMin: Float
    let yMin: Float
    let xMax: Float
    let yMax: Float
    let confidence: Float
    let medianDepthM: Float
    let minDepthM: Float
}

enum RustCoreBridge {
    static func abiVersionHex() -> String {
        let value = apollos_abi_version_u32()
        return String(format: "0x%08X", value)
    }

    static func analyzeDefaultWalkingFrame() -> IOSKinematicResult {
        let output = apollos_analyze_kinematics(
            2,
            1,
            11.0,
            2.0,
            6.0,
            0.0,
            9.8,
            0.2,
            4.0,
            2.0,
            1.0,
            0
        )

        return IOSKinematicResult(
            riskScore: output.risk_score,
            shouldCapture: output.should_capture != 0,
            yawDeltaDeg: output.yaw_delta_deg
        )
    }

    static func depthOnnxEnabled() -> Bool {
        apollos_depth_onnx_runtime_enabled() != 0
    }

    static func eskfCreate() -> UInt64 {
        apollos_eskf_create()
    }

    @discardableResult
    static func eskfDestroy(handle: UInt64) -> Bool {
        if handle == 0 {
            return false
        }
        return apollos_eskf_destroy(handle) != 0
    }

    @discardableResult
    static func eskfReset(handle: UInt64) -> Bool {
        if handle == 0 {
            return false
        }
        return apollos_eskf_reset(handle) != 0
    }

    @discardableResult
    static func eskfPredictImu(handle: UInt64, accelX: Float, accelY: Float, accelZ: Float, dtS: Float) -> Bool {
        if handle == 0 {
            return false
        }
        return apollos_eskf_predict_imu(handle, accelX, accelY, accelZ, dtS) != 0
    }

    @discardableResult
    static func eskfUpdateVision(
        handle: UInt64,
        positionX: Float,
        positionY: Float,
        positionZ: Float,
        varianceM2: Float
    ) -> Bool {
        if handle == 0 {
            return false
        }
        return apollos_eskf_update_vision(handle, positionX, positionY, positionZ, varianceM2) != 0
    }

    static func eskfSnapshot(handle: UInt64) -> IOSEskfSnapshot {
        if handle == 0 {
            return IOSEskfSnapshot(
                sensorHealthScore: 0.0,
                degraded: true,
                localizationUncertaintyM: 999.0,
                innovationNorm: 10.0,
                covarianceXx: 999.0,
                covarianceYy: 999.0,
                covarianceZz: 999.0
            )
        }

        let snapshot = apollos_eskf_snapshot(handle)
        return IOSEskfSnapshot(
            sensorHealthScore: snapshot.sensor_health_score,
            degraded: snapshot.degraded != 0,
            localizationUncertaintyM: snapshot.localization_uncertainty_m,
            innovationNorm: snapshot.innovation_norm,
            covarianceXx: snapshot.covariance_xx,
            covarianceYy: snapshot.covariance_yy,
            covarianceZz: snapshot.covariance_zz
        )
    }


    static func detectDropAheadObjects(
        objects: [IOSEdgeObjectDetection],
        riskScore: Float,
        carryModeCode: UInt8,
        gyroMagnitude: Float,
        nowMs: UInt64
    ) -> IOSDepthHazardResult {
        if objects.isEmpty {
            return IOSDepthHazardResult(
                detected: false,
                positionX: 0.0,
                confidence: 0.0,
                distanceCode: 0
            )
        }

        let mapped = objects.prefix(32).map { obj in
            ApollosObjectSensorFusionInput(
                bbox: ApollosBoundingBox(
                    label_id: obj.labelId,
                    x_min: obj.xMin,
                    y_min: obj.yMin,
                    x_max: obj.xMax,
                    y_max: obj.yMax,
                    confidence: obj.confidence
                ),
                spatial: ApollosDepthSpatials(
                    median_depth_m: obj.medianDepthM,
                    min_depth_m: obj.minDepthM
                )
            )
        }

        let output = mapped.withUnsafeBufferPointer { buffer -> ApollosDepthHazardOutput in
            apollos_detect_drop_ahead_objects(
                buffer.baseAddress,
                mapped.count,
                riskScore,
                carryModeCode,
                gyroMagnitude,
                nowMs
            )
        }

        return IOSDepthHazardResult(
            detected: output.detected != 0,
            positionX: output.position_x,
            confidence: output.confidence,
            distanceCode: output.distance_code
        )
    }
}

File: native/android/app/src/main/java/com/apollos/nativeapp/RustCoreBridge.kt
package com.apollos.nativeapp

import java.nio.ByteBuffer

data class KinematicResult(
    val riskScore: Float,
    val shouldCapture: Boolean,
    val yawDeltaDeg: Float,
)

data class DepthHazardResult(
    val detected: Boolean,
    val positionX: Float,
    val confidence: Float,
    val distanceCode: Int,
    val distanceM: Float,
    val relativeVelocityMps: Float,
    val timeToCollisionS: Float?,
)

data class EdgeObjectDetection(
    val labelId: Int,
    val xMin: Float,
    val yMin: Float,
    val xMax: Float,
    val yMax: Float,
    val confidence: Float,
    val medianDepthM: Float,
    val minDepthM: Float,
)

data class EskfSnapshot(
    val sensorHealthScore: Float,
    val degraded: Boolean,
    val localizationUncertaintyM: Float,
    val innovationNorm: Float,
    val covarianceXx: Float,
    val covarianceYy: Float,
    val covarianceZz: Float,
)

data class VisionOdometryResult(
    val applied: Boolean,
    val deltaXM: Float,
    val deltaYM: Float,
    val poseXM: Float,
    val poseYM: Float,
    val varianceM2: Float,
    val opticalFlowScore: Float,
    val lateralBias: Float,
)

object RustCoreBridge {
    init {
        System.loadLibrary("rust_bridge")
    }

    private external fun nativeAbiVersion(): Int

    private external fun nativeAnalyzeKinematics(
        motionStateCode: Byte,
        carryModeCode: Byte,
        pitch: Float,
        velocity: Float,
        yawDeltaDeg: Float,
        accelX: Float,
        accelY: Float,
        accelZ: Float,
        gyroAlpha: Float,
        gyroBeta: Float,
        gyroGamma: Float,
        sensorUnavailable: Byte,
    ): FloatArray

    private external fun nativeComputeYawDelta(
        alphaDegPerSecond: Float,
        dtMs: Float,
    ): Float

    private external fun nativeDepthOnnxEnabled(): Int


    private external fun nativeDetectDropAheadObjects(
        objectVector: FloatArray,
        objectCount: Int,
        riskScore: Float,
        carryModeCode: Byte,
        gyroMagnitude: Float,
        nowMs: Long,
    ): FloatArray
    private external fun nativeEskfCreate(): Long
    private external fun nativeEskfDestroy(handle: Long): Int
    private external fun nativeEskfReset(handle: Long): Int
    private external fun nativeEskfPredictImu(
        handle: Long,
        accelX: Float,
        accelY: Float,
        accelZ: Float,
        dtS: Float,
    ): Int

    private external fun nativeEskfUpdateVision(
        handle: Long,
        positionX: Float,
        positionY: Float,
        positionZ: Float,
        varianceM2: Float,
    ): Int


    private external fun nativeEskfSnapshot(handle: Long): FloatArray

    fun abiVersion(): Int = nativeAbiVersion()
    fun depthOnnxEnabled(): Boolean = nativeDepthOnnxEnabled() != 0
    fun eskfCreate(): Long = nativeEskfCreate()
    fun eskfDestroy(handle: Long): Boolean = handle != 0L && nativeEskfDestroy(handle) != 0
    fun eskfReset(handle: Long): Boolean = handle != 0L && nativeEskfReset(handle) != 0

    fun eskfPredictImu(handle: Long, accelX: Float, accelY: Float, accelZ: Float, dtS: Float): Boolean {
        if (handle == 0L) {
            return false
        }
        return nativeEskfPredictImu(handle, accelX, accelY, accelZ, dtS) != 0
    }

    fun eskfUpdateVision(
        handle: Long,
        positionX: Float,
        positionY: Float,
        positionZ: Float,
        varianceM2: Float,
    ): Boolean {
        if (handle == 0L) {
            return false
        }
        return nativeEskfUpdateVision(handle, positionX, positionY, positionZ, varianceM2) != 0
    }

    fun eskfSnapshot(handle: Long): EskfSnapshot {
        if (handle == 0L) {
            return EskfSnapshot(
                sensorHealthScore = 0.0f,
                degraded = true,
                localizationUncertaintyM = 999.0f,
                innovationNorm = 10.0f,
                covarianceXx = 999.0f,
                covarianceYy = 999.0f,
                covarianceZz = 999.0f,
            )
        }
        val output = nativeEskfSnapshot(handle)
        return EskfSnapshot(
            sensorHealthScore = output.getOrElse(0) { 0.0f },
            degraded = output.getOrElse(1) { 1.0f } > 0.5f,
            localizationUncertaintyM = output.getOrElse(2) { 999.0f },
            innovationNorm = output.getOrElse(3) { 10.0f },
            covarianceXx = output.getOrElse(4) { 999.0f },
            covarianceYy = output.getOrElse(5) { 999.0f },
            covarianceZz = output.getOrElse(6) { 999.0f },
        )
    }


    fun analyzeKinematics(
        motionStateCode: Byte,
        carryModeCode: Byte,
        pitch: Float,
        velocity: Float,
        yawDeltaDeg: Float,
        accelX: Float,
        accelY: Float,
        accelZ: Float,
        gyroAlpha: Float,
        gyroBeta: Float,
        gyroGamma: Float,
        sensorUnavailable: Boolean,
    ): KinematicResult {
        val output = nativeAnalyzeKinematics(
            motionStateCode,
            carryModeCode,
            pitch,
            velocity,
            yawDeltaDeg,
            accelX,
            accelY,
            accelZ,
            gyroAlpha,
            gyroBeta,
            gyroGamma,
            if (sensorUnavailable) 1 else 0,
        )

        return KinematicResult(
            riskScore = output.getOrElse(0) { 0.0f },
            shouldCapture = output.getOrElse(1) { 0.0f } > 0.5f,
            yawDeltaDeg = output.getOrElse(2) { 0.0f },
        )
    }

    fun computeYawDelta(alphaDegPerSecond: Float, dtMs: Float): Float {
        return nativeComputeYawDelta(alphaDegPerSecond, dtMs)
    }


    fun detectDropAheadObjects(
        objects: List<EdgeObjectDetection>,
        riskScore: Float,
        carryModeCode: Byte,
        gyroMagnitude: Float,
        nowMs: Long,
    ): DepthHazardResult {
        if (objects.isEmpty()) {
            return DepthHazardResult(
                detected = false,
                positionX = 0.0f,
                confidence = 0.0f,
                distanceCode = 0,
                distanceM = 0.0f,
                relativeVelocityMps = 0.0f,
                timeToCollisionS = null,
            )
        }

        val count = objects.size.coerceAtMost(32)
        val packed = FloatArray(count * 8)
        for (idx in 0 until count) {
            val obj = objects[idx]
            val base = idx * 8
            packed[base] = obj.labelId.toFloat().coerceAtLeast(0.0f)
            packed[base + 1] = obj.xMin.coerceIn(0.0f, 1.0f)
            packed[base + 2] = obj.yMin.coerceIn(0.0f, 1.0f)
            packed[base + 3] = obj.xMax.coerceIn(0.0f, 1.0f)
            packed[base + 4] = obj.yMax.coerceIn(0.0f, 1.0f)
            packed[base + 5] = obj.confidence.coerceIn(0.0f, 1.0f)
            packed[base + 6] = obj.medianDepthM.coerceAtLeast(0.0f)
            packed[base + 7] = obj.minDepthM.coerceAtLeast(0.0f)
        }

        val output = nativeDetectDropAheadObjects(
            objectVector = packed,
            objectCount = count,
            riskScore = riskScore,
            carryModeCode = carryModeCode,
            gyroMagnitude = gyroMagnitude,
            nowMs = nowMs,
        )
        return DepthHazardResult(
            detected = output.getOrElse(0) { 0.0f } > 0.5f,
            positionX = output.getOrElse(1) { 0.0f },
            confidence = output.getOrElse(2) { 0.0f },
            distanceCode = output.getOrElse(3) { 0.0f }.toInt(),
            distanceM = output.getOrElse(4) { 0.0f }.coerceAtLeast(0.0f),
            relativeVelocityMps = output.getOrElse(5) { 0.0f },
            timeToCollisionS = output.getOrNull(6)?.takeIf { it.isFinite() && it > 0.0f },
        )
    }
}
